### DISCUSSION TITLE
Minimal state to ensure reproducible simulations?

### DISCUSSION THREAD

**Discussion by 7x11x13** [JSBSim-Team/jsbsim#942] [Q&A]
I'm working on a project in Unreal Engine using JSBSim and I'm attempting to get it to work with multiple players. It's working fine, however in the case of a client/server desync, I'm not sure exactly what information I need to send the client from the server to correct its local simulation (and how best to do it). Is setting the position/rotation/velocity enough for the simulations to sync up again (assuming no wind)? And if so what is the best way to do that? Can I modify these values mid simulation or should I reset entirely with new initial conditions?

**Comment by seanmcleod** [comment]
Just to clarify, assuming say 2 players, are they both operating the same aircraft, e.g. pilot and co-pilot and so need to see the exact same state on their PCs? Or are they flying their own individual aircraft? 

If they're flying their own aircraft, do you need to transmit more than the aircraft's pose (position and attitude), so that you can render the other player's aircraft in the virtual world correctly?

**Reply by 7x11x13** [reply]
They are flying their own individual aircraft. I have one simulation running on the server and one on the client for each player's aircraft, which is why I need the simulations to sync up. This is done to prevent cheating (i.e. if the simulation was just being run on the client's machine and transmitting position data to the server, the client could send bogus position data) while still being responsive on the client's end (so they don't have to wait for a response from the server every time they change their controls). I'm overriding Unreal's CharacterMovementComponent to provide this functionality.

**Comment by seanmcleod** [comment]
How are you keeping things in sync currently assuming no network issues? Do you forward, at 120Hz (default JSBSim sim rate), all the control inputs on PC-A which are fed into an instance of JSBSim on PC-A to simulate Aircraft-A, to PC-B via the server, where PC-B then feeds these control inputs into it's Aircraft-A JSBSim instance that it has running to mirror Aircraft-A?

**Reply by 7x11x13** [reply]
That is correct, where PC-A is the client and PC-B is the server. If there are two players involved then there would be also a third PC-C which sends its own control inputs for Aircraft-C to PC-B and also receives just position info (not control inputs) from PC-B for displaying Aircraft-A.

**Comment by seanmcleod** [comment]
So if there is say a 2s break in comms between PC-A and the server, one option would be to send a list/queue of the control inputs that were made by the user to the server, where the server could feed them into the JSBSim instance it's running to mirror Aircraft-A, faster than real-time to catch up.

**Reply by 7x11x13** [reply]
Right, I believe CharacterMovementComponent is doing this already. However, suppose the player on PC-A makes some control inputs in this 2s break. Then the simulation on the server will be as if PC-A never made any new control inputs, so it will have to "rewind" back 2s before it replays these inputs (alternatively, you could have PC-A get corrected to the current state of the server, which is what I was asking in my initial post). Either way, there must be some mechanism to change the state of the simulation on one of the computers.

**Comment by seanmcleod** [comment]
I was thinking that the server in this case would pause simulation, no control input packet no simulation timestep for the particular aircraft, then when it received 2mins worth of inputs when the network connection was restored and play the 2min worth of inputs in a fraction of a second.

> must be some mechanism to change the state of the simulation 

And what level of state, the full state including things like fuel state etc.?

**Reply by 7x11x13** [reply]
Currently I'm doing the following to set the state. It seems to work fine at very low speeds but it gets jittery at higher speeds, so I think it's not completely syncing up the simulations.
```
IC->InitializeIC();

IC->SetGeodLatitudeDegIC(container.Location.X);
IC->SetLongitudeDegIC(container.Location.Y);
IC->SetAltitudeASLFtIC(container.Location.Z);

IC->SetPsiDegIC(container.LocalEulerAngles.Yaw);
IC->SetThetaDegIC(container.LocalEulerAngles.Pitch);
IC->SetPhiDegIC(container.LocalEulerAngles.Roll);

IC->SetPRadpsIC(container.EulerRates.X);
IC->SetQRadpsIC(container.EulerRates.Y);
IC->SetRRadpsIC(container.EulerRates.Z);

IC->SetVNorthFpsIC(container.VelocityNEDfps.X);
IC->SetVEastFpsIC(container.VelocityNEDfps.Y);
IC->SetVDownFpsIC(-container.VelocityNEDfps.Z);

Propagate->SetInitialState(IC.get());
```

**Comment by seanmcleod** [comment]
What about all the other states? There are potentially a large number of states that you haven't included, here are just a couple off the top of my head.

- Control inputs, including things like gear, flaps etc.
- Control positions, not always 1:1 with control inputs, e.g. there can be actuators with lag, rates, delays etc. 
- Engine state, again generally not simply 1:1 with throttle input.
- Aircraft mass. If the snapshot you're retrieving your "initial conditions" from has been running for some time this isn't the same as the initial mass.
- Fuel tank states. What order of tanks have been used up until you took an "initial conditions" snapshot?
- Moments of inertia, see aircraft mass, as fuel is burnt off, given location of fuel tank being used etc. this will result in a different moment of inertia compared to the initial moment of inertia.

Now if in your initial testing you've only had the first instance running for a couple of seconds, then took a snapshot, and then launched a new JSBSim instance, or update an existing instance that has been out of sync for a while and tried to initialize it the differences I mentioned above may not be that big a deal. But if there is a larger time period there will be.
